File: GdPicture/WinForm/vb.net/Document Clean Up/Effects.vb
Option Explicit On
Option Infer On
Option Strict On

Imports GdPicture14

Namespace DocumentCleanUp

    ' -------------------------------
    ' class definition
    ' -------------------------------

    Public MustInherit Class ADocumentCleanUpEffectParameters
        Public Function Clone() As ADocumentCleanUpEffectParameters
            Return DirectCast(Me.MemberwiseClone(), ADocumentCleanUpEffectParameters)
        End Function
    End Class

    Public MustInherit Class ADocumentCleanUpEffect
        Protected _parameters As ADocumentCleanUpEffectParameters = Nothing
        Public MustOverride ReadOnly Property Name() As String

        Public ReadOnly Property Parameters() As ADocumentCleanUpEffectParameters
            Get
                Return _parameters
            End Get
        End Property

        Public Function Clone() As ADocumentCleanUpEffect
            Dim obj As ADocumentCleanUpEffect = DirectCast(Me.MemberwiseClone(), ADocumentCleanUpEffect)
            If Not obj._parameters Is Nothing Then
                obj._parameters = Me._parameters.Clone()
            End If
            Return obj
        End Function
    End Class

    ' -------------------------------
    ' effects class definition
    ' -------------------------------

    ' Auto deskew
    Public Class AutoDeskewEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property MaxAngleOfResearch As Single = 10
        Public Property BackColor As Color = Color.White
        Public Property AngleStep As Single = 0.25
        Public Property Optimistic As Boolean = True
    End Class

    Public Class AutoDeskewEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New AutoDeskewEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Auto deskew"
            End Get
        End Property
    End Class

    ' Auto invert negative
    Public Class AutoInvertEffect
        Inherits ADocumentCleanUpEffect

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Auto invert negative"
            End Get
        End Property
    End Class

    ' Auto text invert
    Public Class AutoTextInvertEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property MinObjectWidth As Integer
        Public Property MinObjectHeight As Integer
        Public Property MinBlackPercent As Integer
        Public Property MaxBlackPercent As Integer
        Public Property FixBackground As Boolean
        Public Property ConnectedToLines As Boolean
        Public Property MaxLinesThickness As Integer
    End Class

    Public Class AutoTextInvertEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New AutoTextInvertEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Auto text invert"
            End Get
        End Property
    End Class

    ' Crop black borders
    Public Class CropBlackBordersEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Confidence As Single = 75
        Public Property SkipLinesCount As Integer
    End Class

    Public Class CropBlackBordersEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New CropBlackBordersEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Crop black borders"
            End Get
        End Property
    End Class

    ' Crop black borders
    Public Class CropBlackBordersEffectExParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Confidence As Single = 75
        Public Property SkipLinesCount As Integer
    End Class

    Public Class CropBlackBordersExEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New CropBlackBordersEffectExParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Crop black borders extended"
            End Get
        End Property
    End Class

    ' Crop black borders
    Public Class CropBordersEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Confidence As Single = 75
        Public Property PixelReference As ReferencePoint
        Public Property Left As Integer
        Public Property Top As Integer
        Public Property Width As Integer
        Public Property Height As Integer
        Public Property Context As ImagingContext
    End Class

    Public Class CropBordersEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New CropBordersEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Crop borders"
            End Get
        End Property
    End Class

    ' Crop white borders
    Public Class CropWhiteBordersEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Confidence As Single = 99
        Public Property SkipLinesCount As Integer
    End Class

    Public Class CropWhiteBordersEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New CropWhiteBordersEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Crop white borders"
            End Get
        End Property
    End Class

    ' delete black borders
    Public Class DeleteBlackBordersEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Margin As Integer = 0
        Public Property SkewedBorders As Boolean = False
    End Class

    Public Class DeleteBlackBordersEffect
        Inherits ADocumentCleanUpEffect

        Public Sub New()
            _parameters = New DeleteBlackBordersEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Delete black borders"
            End Get
        End Property
    End Class

    ' despeckle
    Public Class DespeckleEffect
        Inherits ADocumentCleanUpEffect

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Despeckle"
            End Get
        End Property
    End Class

    ' despeckle more
    Public Class DespeckleMoreEffect
        Inherits ADocumentCleanUpEffect

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Despeckle more"
            End Get
        End Property
    End Class

    ' Despeckle advanced
    Public Class DespeckAdvancedParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property DespeckleMode As DespeckleType = DespeckleType.FourConnected
        Public Property Strength As Integer = 5
    End Class

    Public Class DespeckleAdvancedEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New DespeckAdvancedParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Despeckle (advanced)"
            End Get
        End Property
    End Class

    ' median
    Public Class MedianEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property KernelSize As Integer = 1
    End Class

    Public Class MedianEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New MedianEffectParameters()
        End Sub

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Median"
            End Get
        End Property
    End Class

    ' remove blob
    Public Class RemoveBlobEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property MinBlobWidth As Integer = 20
        Public Property MinBlobHeight As Integer = 20
        Public Property MaxBlobWidth As Integer = 100
        Public Property MaxBlobHeight As Integer = 100
        Public Property MinFillPercent As Integer = 75
        Public Property MaxFillPercent As Integer = 100
    End Class

    Public Class RemoveBlobEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New RemoveBlobEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Remove blob"
            End Get
        End Property
    End Class

    ' remove punch holes
    Public Class RemoveHolePunchEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Margins As HolePunchMargins = HolePunchMargins.MarginBottom Or HolePunchMargins.MarginLeft Or HolePunchMargins.MarginRight Or HolePunchMargins.MarginTop
    End Class

    Public Class RemoveHolePunchEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New RemoveHolePunchEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Remove punch holes"
            End Get
        End Property
    End Class

    ' remove lines
    Public Class RemoveLinesEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Orientation As LineRemoveOrientation = LineRemoveOrientation.Horizontal
        Public Property MaxLineGap As Integer = -1
        Public Property MaxLineThickness As Integer = 8
        Public Property MinLineLength As Integer = 75
        Public Property MaxInterception As Integer = 8
        Public Property ReConnectBrokenCharacters As Boolean = False
    End Class

    Public Class RemoveLinesEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New RemoveLinesEffectParameters
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Remove lines"
            End Get
        End Property
    End Class

    'remove staple mark
    Public Class RemoveStapleMarkEffect
        Inherits ADocumentCleanUpEffect

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Remove staple mark"
            End Get
        End Property
    End Class

    ' bitonal close8
    Public Class BitonalClose8Effect
        Inherits ADocumentCleanUpEffect

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal morphological close"
            End Get
        End Property
    End Class

    ' bitonal despeckle
    Public Class BitonalDespeckleEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property FixText As Boolean
    End Class

    Public Class BitonalDespeckleEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New BitonalDespeckleEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal despeckle"
            End Get
        End Property
    End Class

    ' bitonal despeckle more
    Public Class BitonalDespeckleMoreEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property FixText As Boolean
    End Class

    Public Class BitonalDespeckleMoreEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New BitonalDespeckleMoreEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal despeckle more"
            End Get
        End Property
    End Class

    ' bitonal ouline
    Public Class BitonalOutlineEffect
        Inherits ADocumentCleanUpEffect

        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal outline"
            End Get
        End Property
    End Class

    ' bitonal skeletonize
    Public Class BitonalSkeletonizeEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property Repetition As Integer = 4
    End Class

    Public Class BitonalSkeletonizeEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New BitonalSkeletonizeEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal skeletonize"
            End Get
        End Property
    End Class

    ' bitonal vigorous despeckle
    Public Class BitonalVigorousDespeckleEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property CheckForDotOfI As Boolean = False
    End Class

    Public Class BitonalVigorousDespeckleEffect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New BitonalVigorousDespeckleEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal vigorous despeckle"
            End Get
        End Property
    End Class

    ' bitonal dilate
    Public Class BitonalDilateEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property KernelSize As Integer = 1
    End Class

    Public Class BitonalDilateEFfect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New BitonalDilateEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal dilate"
            End Get
        End Property
    End Class

    ' bitonal dilate 4
    Public Class BitonalDilate4EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal dilate 4 directions"
            End Get
        End Property
    End Class

    ' bitonal dilate 8
    Public Class BitonalDilate8EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal dilate 8 directions"
            End Get
        End Property
    End Class

    ' bitonal dilate h
    Public Class BitonalDilateHEFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal dilate horizontal"
            End Get
        End Property
    End Class

    ' bitonal dilate v
    Public Class BitonalDilateVEFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal dilate vertical"
            End Get
        End Property
    End Class

    ' bitonal erode
    Public Class BitonalErodeEffectParameters
        Inherits ADocumentCleanUpEffectParameters
        Public Property KernelSize As Integer = 1
    End Class

    Public Class BitonalErodeEFfect
        Inherits ADocumentCleanUpEffect
        Public Sub New()
            _parameters = New BitonalErodeEffectParameters()
        End Sub
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal erode"
            End Get
        End Property
    End Class

    ' bitonal erode 4
    Public Class BitonalErode4EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal erode 4 directions"
            End Get
        End Property
    End Class

    ' bitonal erode 8
    Public Class BitonalErode8EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal erode 8 directions"
            End Get
        End Property
    End Class

    ' bitonal erode v
    Public Class BitonalErodeVEFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal erode vertical"
            End Get
        End Property
    End Class

    ' bitonal fill holes horizontal
    Public Class BitonalFillHolesHEFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal fill holes horizontal"
            End Get
        End Property
    End Class

    ' bitonal fill holes horizontal & vertical
    Public Class BitonalFillHolesHVEFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal fill holes horizontal and vertical"
            End Get
        End Property
    End Class

    ' bitonal fill holes vertical
    Public Class BitonalFillHolesVEFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal fill holes vertical"
            End Get
        End Property
    End Class

    ' bitonal open 8
    Public Class BitonalOpen8EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal morphological open"
            End Get
        End Property
    End Class

    ' bitonal remove isolated dots 2x2
    Public Class BitonalRemoveIsolatedDots2x2EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal remove isolated dots 2x2"
            End Get
        End Property
    End Class

    ' bitonal remove isolated dots 4
    Public Class BitonalRemoveIsolatedDots4EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal remove isolated dots 4 directions"
            End Get
        End Property
    End Class

    ' bitonal remove isolated dots 8
    Public Class BitonalRemoveIsolatedDots8EFfect
        Inherits ADocumentCleanUpEffect
        Public Overrides ReadOnly Property Name As String
            Get
                Return "Bitonal remove isolated dots 8 directions"
            End Get
        End Property
    End Class

    ' -------------------------------
    ' effects list
    ' -------------------------------
    Public Class DocumentCleanUpEffects
        Private ReadOnly _list As List(Of ADocumentCleanUpEffect) = New List(Of ADocumentCleanUpEffect)()

        Public Sub New()
            _list.Add(New AutoDeskewEffect())
            _list.Add(New AutoInvertEffect())
            _list.Add(New AutoTextInvertEffect())
            _list.Add(New CropBlackBordersEffect())
            _list.Add(New CropBlackBordersExEffect())
            _list.Add(New CropBordersEffect)
            _list.Add(New CropWhiteBordersEffect)
            _list.Add(New DeleteBlackBordersEffect())
            _list.Add(New DespeckleEffect())
            _list.Add(New DespeckleMoreEffect())
            _list.Add(New DespeckleAdvancedEffect())
            _list.Add(New MedianEffect())
            _list.Add(New RemoveBlobEffect())
            _list.Add(New RemoveHolePunchEffect())
            _list.Add(New RemoveLinesEffect())
            _list.Add(New RemoveStapleMarkEffect())
            _list.Add(New BitonalClose8Effect())
            _list.Add(New BitonalDespeckleEffect())
            _list.Add(New BitonalDespeckleMoreEffect())
            _list.Add(New BitonalOutlineEffect())
            _list.Add(New BitonalSkeletonizeEffect())
            _list.Add(New BitonalVigorousDespeckleEffect())
            _list.Add(New BitonalDilateEFfect())
            _list.Add(New BitonalDilate4EFfect())
            _list.Add(New BitonalDilate8EFfect())
            _list.Add(New BitonalDilateHEFfect())
            _list.Add(New BitonalDilateVEFfect())
            _list.Add(New BitonalErodeEFfect())
            _list.Add(New BitonalErode4EFfect())
            _list.Add(New BitonalErode8EFfect())
            _list.Add(New BitonalErodeVEFfect())
            _list.Add(New BitonalFillHolesHEFfect())
            _list.Add(New BitonalFillHolesHVEFfect())
            _list.Add(New BitonalFillHolesVEFfect())
            _list.Add(New BitonalOpen8EFfect())
            _list.Add(New BitonalRemoveIsolatedDots2x2EFfect())
            _list.Add(New BitonalRemoveIsolatedDots4EFfect())
            _list.Add(New BitonalRemoveIsolatedDots8EFfect())
        End Sub

        Public ReadOnly Property List() As List(Of ADocumentCleanUpEffect)
            Get
                Return _list
            End Get
        End Property
    End Class
End Namespace

File: GdPicture/WinForm/C#/Document Clean Up/Effects.cs
using GdPicture14;
using System.Collections.Generic;
using System.Drawing;



namespace DocumentCleanUp
{
    namespace DocumentCleanUp
    {

        // -------------------------------
        // class definition
        // -------------------------------

        public abstract class ADocumentCleanUpEffectParameters
        {
            public ADocumentCleanUpEffectParameters Clone()
            {
                return ((ADocumentCleanUpEffectParameters)(this.MemberwiseClone()));
            }
        }

        public abstract class ADocumentCleanUpEffect
        {
            protected ADocumentCleanUpEffectParameters _parameters = null;
            public abstract string Name { get; }

            public ADocumentCleanUpEffectParameters Parameters
            {
                get
                {
                    return _parameters;
                }
            }

            public ADocumentCleanUpEffect Clone()
            {
                ADocumentCleanUpEffect obj = (ADocumentCleanUpEffect)(this.MemberwiseClone());
                if (obj._parameters != null)
                {
                    obj._parameters = this._parameters.Clone();
                }
                return obj;
            }
        }

        // -------------------------------
        // effects class definition
        // -------------------------------

        // Auto deskew
        public class AutoDeskewEffectParameters : ADocumentCleanUpEffectParameters
        {
            public AutoDeskewEffectParameters()
            {
                // VBConversions Note: Non-static class variable initialization is below.  Class variables cannot be initially assigned non-static values in C#.
                BackColor = Color.White;

            }
            public float MaxAngleOfResearch { get; set; } = 10;
            public Color BackColor { get; set; }
            public float AngleStep { get; set; } = 0.25F;
            public bool Optimistic { get; set; } = true;
        }

        public class AutoDeskewEffect : ADocumentCleanUpEffect
        {
            public AutoDeskewEffect()
            {
                _parameters = new AutoDeskewEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Auto deskew";
                }
            }
        }

        // Auto invert negative
        public class AutoInvertEffect : ADocumentCleanUpEffect
        {

            public override string Name
            {
                get
                {
                    return "Auto invert negative";
                }
            }
        }

        // Auto text invert
        public class AutoTextInvertEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int MinObjectWidth { get; set; }
            public int MinObjectHeight { get; set; }
            public int MinBlackPercent { get; set; }
            public int MaxBlackPercent { get; set; }
            public bool FixBackground { get; set; }
            public bool ConnectedToLines { get; set; }
            public int MaxLinesThickness { get; set; }
        }

        public class AutoTextInvertEffect : ADocumentCleanUpEffect
        {
            public AutoTextInvertEffect()
            {
                _parameters = new AutoTextInvertEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Auto text invert";
                }
            }
        }

        // Crop black borders
        public class CropBlackBordersEffectParameters : ADocumentCleanUpEffectParameters
        {
            public float Confidence { get; set; } = 75;
            public int SkipLinesCount { get; set; }
        }

        public class CropBlackBordersEffect : ADocumentCleanUpEffect
        {
            public CropBlackBordersEffect()
            {
                _parameters = new CropBlackBordersEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Crop black borders";
                }
            }
        }

        // Crop black borders
        public class CropBlackBordersEffectExParameters : ADocumentCleanUpEffectParameters
        {
            public float Confidence { get; set; } = 75;
            public int SkipLinesCount { get; set; }
        }

        public class CropBlackBordersExEffect : ADocumentCleanUpEffect
        {
            public CropBlackBordersExEffect()
            {
                _parameters = new CropBlackBordersEffectExParameters();
            }

            public override string Name
            {
                get
                {
                    return "Crop black borders extended";
                }
            }
        }

        // Crop borders
        public class CropBordersEffectParameters : ADocumentCleanUpEffectParameters
        {
            public float Confidence { get; set; } = 99;
            public ReferencePoint PixelReference { get; set; }
            public int Left { get; set; }
            public int Top { get; set; }
            public int Width { get; set; }
            public int Height { get; set; }
            public ImagingContext Context { get; set; }
        }

        public class CropBordersEffect : ADocumentCleanUpEffect
        {
            public CropBordersEffect()
            {
                _parameters = new CropBordersEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Crop borders";
                }
            }
        }

        // Crop white borders
        public class CropWhiteBordersEffectParameters : ADocumentCleanUpEffectParameters
        {
            public float Confidence { get; set; } = 99;
            public int SkipLinesCount { get; set; }
        }

        public class CropWhiteBordersEffect : ADocumentCleanUpEffect
        {
            public CropWhiteBordersEffect()
            {
                _parameters = new CropWhiteBordersEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Crop white borders";
                }
            }
        }

        // delete black borders
        public class DeleteBlackBordersEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int Margin { get; set; } = 0;
            public bool SkewedBorders { get; set; } = false;
        }

        public class DeleteBlackBordersEffect : ADocumentCleanUpEffect
        {

            public DeleteBlackBordersEffect()
            {
                _parameters = new DeleteBlackBordersEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Delete black borders";
                }
            }
        }

        // despeckle
        public class DespeckleEffect : ADocumentCleanUpEffect
        {

            public override string Name
            {
                get
                {
                    return "Despeckle";
                }
            }
        }

        // despeckle more
        public class DespeckleMoreEffect : ADocumentCleanUpEffect
        {

            public override string Name
            {
                get
                {
                    return "Despeckle more";
                }
            }
        }

        // Despeckle advanced
        public class DespeckAdvancedParameters : ADocumentCleanUpEffectParameters
        {
            public DespeckleType DespeckleMode { get; set; } = DespeckleType.FourConnected;
            public int Strength { get; set; } = 5;
        }

        public class DespeckleAdvancedEffect : ADocumentCleanUpEffect
        {
            public DespeckleAdvancedEffect()
            {
                _parameters = new DespeckAdvancedParameters();
            }

            public override string Name
            {
                get
                {
                    return "Despeckle (advanced)";
                }
            }
        }

        // median
        public class MedianEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int KernelSize { get; set; } = 1;
        }

        public class MedianEffect : ADocumentCleanUpEffect
        {
            public MedianEffect()
            {
                _parameters = new MedianEffectParameters();
            }

            public override string Name
            {
                get
                {
                    return "Median";
                }
            }
        }

        // remove blob
        public class RemoveBlobEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int MinBlobWidth { get; set; } = 20;
            public int MinBlobHeight { get; set; } = 20;
            public int MaxBlobWidth { get; set; } = 100;
            public int MaxBlobHeight { get; set; } = 100;
            public int MinFillPercent { get; set; } = 75;
            public int MaxFillPercent { get; set; } = 100;
        }

        public class RemoveBlobEffect : ADocumentCleanUpEffect
        {
            public RemoveBlobEffect()
            {
                _parameters = new RemoveBlobEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Remove blob";
                }
            }
        }

        // remove punch holes
        public class RemoveHolePunchEffectParameters : ADocumentCleanUpEffectParameters
        {
            public HolePunchMargins Margins { get; set; } = HolePunchMargins.MarginLeft | HolePunchMargins.MarginRight | HolePunchMargins.MarginBottom | HolePunchMargins.MarginTop;
        }

        public class RemoveHolePunchEffect : ADocumentCleanUpEffect
        {
            public RemoveHolePunchEffect()
            {
                _parameters = new RemoveHolePunchEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Remove punch holes";
                }
            }
        }

        // remove lines
        public class RemoveLinesEffectParameters : ADocumentCleanUpEffectParameters
        {
            public LineRemoveOrientation Orientation { get; set; } = LineRemoveOrientation.Horizontal;
            public int MaxLineGap { get; set; } = -1;
            public int MaxLineThickness { get; set; } = 8;
            public int MinLineLength { get; set; } = 75;
            public int MaxInterception { get; set; } = 8;
            public bool ReConnectBrokenCharacters { get; set; } = false;
        }

        public class RemoveLinesEffect : ADocumentCleanUpEffect
        {
            public RemoveLinesEffect()
            {
                _parameters = new RemoveLinesEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Remove lines";
                }
            }
        }

        //remove staple mark
        public class RemoveStapleMarkEffect : ADocumentCleanUpEffect
        {

            public override string Name
            {
                get
                {
                    return "Remove staple mark";
                }
            }
        }

        // bitonal close8
        public class BitonalClose8Effect : ADocumentCleanUpEffect
        {

            public override string Name
            {
                get
                {
                    return "Bitonal morphological close";
                }
            }
        }

        // bitonal despeckle
        public class BitonalDespeckleEffectParameters : ADocumentCleanUpEffectParameters
        {
            public bool FixText { get; set; }
        }

        public class BitonalDespeckleEffect : ADocumentCleanUpEffect
        {
            public BitonalDespeckleEffect()
            {
                _parameters = new BitonalDespeckleEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Bitonal despeckle";
                }
            }
        }

        // bitonal despeckle more
        public class BitonalDespeckleMoreEffectParameters : ADocumentCleanUpEffectParameters
        {
            public bool FixText { get; set; }
        }

        public class BitonalDespeckleMoreEffect : ADocumentCleanUpEffect
        {
            public BitonalDespeckleMoreEffect()
            {
                _parameters = new BitonalDespeckleMoreEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Bitonal despeckle more";
                }
            }
        }

        // bitonal ouline
        public class BitonalOutlineEffect : ADocumentCleanUpEffect
        {

            public override string Name
            {
                get
                {
                    return "Bitonal outline";
                }
            }
        }

        // bitonal skeletonize
        public class BitonalSkeletonizeEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int Repetition { get; set; } = 4;
        }

        public class BitonalSkeletonizeEffect : ADocumentCleanUpEffect
        {
            public BitonalSkeletonizeEffect()
            {
                _parameters = new BitonalSkeletonizeEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Bitonal skeletonize";
                }
            }
        }

        // bitonal vigorous despeckle
        public class BitonalVigorousDespeckleEffectParameters : ADocumentCleanUpEffectParameters
        {
            public bool CheckForDotOfI { get; set; } = false;
        }

        public class BitonalVigorousDespeckleEffect : ADocumentCleanUpEffect
        {
            public BitonalVigorousDespeckleEffect()
            {
                _parameters = new BitonalVigorousDespeckleEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Bitonal vigorous despeckle";
                }
            }
        }

        // bitonal dilate
        public class BitonalDilateEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int KernelSize { get; set; } = 1;
        }

        public class BitonalDilateEffect : ADocumentCleanUpEffect
        {
            public BitonalDilateEffect()
            {
                _parameters = new BitonalDilateEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Bitonal dilate";
                }
            }
        }

        // bitonal dilate 4
        public class BitonalDilate4Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal dilate 4 directions";
                }
            }
        }

        // bitonal dilate 8
        public class BitonalDilate8Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal dilate 8 directions";
                }
            }
        }

        // bitonal dilate h
        public class BitonalDilateHEffect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal dilate horizontal";
                }
            }
        }

        // bitonal dilate v
        public class BitonalDilateVEffect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal dilate vertical";
                }
            }
        }

        // bitonal erode
        public class BitonalErodeEffectParameters : ADocumentCleanUpEffectParameters
        {
            public int KernelSize { get; set; } = 1;
        }

        public class BitonalErodeEffect : ADocumentCleanUpEffect
        {
            public BitonalErodeEffect()
            {
                _parameters = new BitonalErodeEffectParameters();
            }
            public override string Name
            {
                get
                {
                    return "Bitonal erode";
                }
            }
        }

        // bitonal erode 4
        public class BitonalErode4Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal erode 4 directions";
                }
            }
        }

        // bitonal erode 8
        public class BitonalErode8Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal erode 8 directions";
                }
            }
        }

        // bitonal erode v
        public class BitonalErodeVEffect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal erode vertical";
                }
            }
        }

        // bitonal fill holes horizontal
        public class BitonalFillHolesHEffect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal fill holes horizontal";
                }
            }
        }

        // bitonal fill holes horizontal & vertical
        public class BitonalFillHolesHVEffect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal fill holes horizontal and vertical";
                }
            }
        }

        // bitonal fill holes vertical
        public class BitonalFillHolesVEffect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal fill holes vertical";
                }
            }
        }

        // bitonal open 8
        public class BitonalOpen8Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal morphological open";
                }
            }
        }

        // bitonal remove isolated dots 2x2
        public class BitonalRemoveIsolatedDots2X2Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal remove isolated dots 2x2";
                }
            }
        }

        // bitonal remove isolated dots 4
        public class BitonalRemoveIsolatedDots4Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal remove isolated dots 4 directions";
                }
            }
        }

        // bitonal remove isolated dots 8
        public class BitonalRemoveIsolatedDots8Effect : ADocumentCleanUpEffect
        {
            public override string Name
            {
                get
                {
                    return "Bitonal remove isolated dots 8 directions";
                }
            }
        }

        // -------------------------------
        // effects list
        // -------------------------------
        public class DocumentCleanUpEffects
        {
            private List<ADocumentCleanUpEffect> _list = new List<ADocumentCleanUpEffect>();

            public DocumentCleanUpEffects()
            {
                _list.Add(new AutoDeskewEffect());
                _list.Add(new AutoInvertEffect());
                _list.Add(new AutoTextInvertEffect());
                _list.Add(new CropBlackBordersEffect());
                _list.Add(new CropBlackBordersExEffect());
                _list.Add(new CropBordersEffect());
                _list.Add(new CropWhiteBordersEffect());
                _list.Add(new DeleteBlackBordersEffect());
                _list.Add(new DespeckleEffect());
                _list.Add(new DespeckleMoreEffect());
                _list.Add(new DespeckleAdvancedEffect());
                _list.Add(new MedianEffect());
                _list.Add(new RemoveBlobEffect());
                _list.Add(new RemoveHolePunchEffect());
                _list.Add(new RemoveLinesEffect());
                _list.Add(new RemoveStapleMarkEffect());
                _list.Add(new BitonalClose8Effect());
                _list.Add(new BitonalDespeckleEffect());
                _list.Add(new BitonalDespeckleMoreEffect());
                _list.Add(new BitonalOutlineEffect());
                _list.Add(new BitonalSkeletonizeEffect());
                _list.Add(new BitonalVigorousDespeckleEffect());
                _list.Add(new BitonalDilateEffect());
                _list.Add(new BitonalDilate4Effect());
                _list.Add(new BitonalDilate8Effect());
                _list.Add(new BitonalDilateHEffect());
                _list.Add(new BitonalDilateVEffect());
                _list.Add(new BitonalErodeEffect());
                _list.Add(new BitonalErode4Effect());
                _list.Add(new BitonalErode8Effect());
                _list.Add(new BitonalErodeVEffect());
                _list.Add(new BitonalFillHolesHEffect());
                _list.Add(new BitonalFillHolesHVEffect());
                _list.Add(new BitonalFillHolesVEffect());
                _list.Add(new BitonalOpen8Effect());
                _list.Add(new BitonalRemoveIsolatedDots2X2Effect());
                _list.Add(new BitonalRemoveIsolatedDots4Effect());
                _list.Add(new BitonalRemoveIsolatedDots8Effect());
            }

            public List<ADocumentCleanUpEffect> List
            {
                get
                {
                    return _list;
                }
            }
        }
    }
}
